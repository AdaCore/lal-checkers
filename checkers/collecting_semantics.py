"""
An analyzer that collects semantics at each program point.
"""

from lalcheck.irs.basic.tools import (
    CFGBuilder,
    ExprEvaluator,
    ExprSolver
)

from lalcheck.irs.basic.tree import Variable
from lalcheck.irs.basic.tools import PrettyPrinter
from lalcheck.irs.basic.purpose import SyntheticVariable
from lalcheck.irs.basic import visitors
from lalcheck.utils import KeyCounter
from lalcheck.digraph import Digraph
from lalcheck import domains
from lalcheck import dot_printer

from xml.sax.saxutils import escape
from collections import defaultdict


class VarTracker(visitors.CFGNodeVisitor):
    def __init__(self, var_set, vars_domain, vars_idx, evaluator, c_solver):
        self.vars = var_set
        self.vars_domain = vars_domain
        self.vars_idx = vars_idx
        self.evaluator = evaluator
        self.constr_solver = c_solver

    def state_to_env(self, state):
        return {v: state[self.vars_idx[v]] for v in self.vars}

    def env_to_state(self, env):
        if len(env) == 0:
            return self.vars_domain.bottom
        else:
            return tuple(env[v] for v in self.vars)

    def visit_assign(self, assign, state):
        env = self.state_to_env(state)
        env[assign.id.var] = self.evaluator.eval(assign.expr, env)
        return self.env_to_state(env)

    def visit_assume(self, assume, state):
        return self.env_to_state(self.constr_solver.solve(
            assume.expr,
            self.state_to_env(state)
        ))

    def visit_read(self, read, state):
        return tuple(self.vars_domain.top[i] if i == self.vars_idx[read.id.var]
                     else x for i, x in enumerate(state))

    def visit_use(self, use, state):
        return state


class SimpleTraceLattice(domains.FiniteSubsetLattice):
    def __init__(self, *args):
        super(SimpleTraceLattice, self).__init__(*args)

    def update(self, a, b, widen=False):
        return super(SimpleTraceLattice, self).update(a, b, False)


class MergePredicateBuilder(object):
    def __init__(self, predicate):
        self.predicate = predicate

    def __or__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) or o_pred(a, b)

        return MergePredicateBuilder(f)

    def __and__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) and o_pred(a, b)

        return MergePredicateBuilder(f)

    def build(self, trace_domain, vals_domain):
        return self.predicate(trace_domain, vals_domain)


def _mp_always(*_):
    return lambda *_: True


def _mp_never(*_):
    return lambda *_: False


def _mp_le_traces(trace_domain, _):
    return lambda a, b: trace_domain.le(a[0], b[0])


def _mp_eq_vals(_, vals_domain):
    return lambda a, b: vals_domain.eq(a[1], b[1])


MergePredicateBuilder.Always = MergePredicateBuilder(_mp_always)
MergePredicateBuilder.Never = MergePredicateBuilder(_mp_never)
MergePredicateBuilder.Le_Traces = MergePredicateBuilder(_mp_le_traces)
MergePredicateBuilder.Eq_Vals = MergePredicateBuilder(_mp_eq_vals)


def html_render_node(node):
    return escape(PrettyPrinter.pretty_print(node))


def save_cfg_to(file_name, cfg):
    def render_node(node):
        return (html_render_node(node),) if node is not None else ()

    def render_widening_point(is_widening_point):
        return (escape('<widening_point>'),) if is_widening_point else ()

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(cfg, [
            dot_printer.DataPrinter('node', render_node),
            dot_printer.DataPrinter('is_widening_point', render_widening_point)
        ]))


def repr_trace(trace):
    return tuple(p.name for p in trace)


def build_resulting_graph(file_name, cfg, results, trace_domain):
    paths = defaultdict(list)

    var_set = {
        v
        for node, state in results.iteritems()
        for trace, values in state.iteritems()
        for v, _ in values.iteritems()
    }

    for node, state in results.iteritems():
        for trace, values in state.iteritems():
            paths[frozenset(trace)].append(
                Digraph.Node(
                    node.name,
                    ___orig=node,
                    **{
                        v.name: value
                        for v, value in values.iteritems()
                        if not SyntheticVariable.is_purpose_of(v)
                    }
                )
            )

    edges = []
    for trace, nodes in paths.iteritems():
        for node in nodes:
            predecessors = [
                n
                for t, ns in paths.iteritems()
                for n in ns
                if trace_domain.le(t, trace)
                if n.data.___orig in cfg.ancestors(node.data.___orig)
            ]

            for pred in predecessors:
                edges.append(Digraph.Edge(pred, node))

    res_graph = Digraph(
        [n for _, nodes in paths.iteritems() for n in nodes],
        edges
    )

    def print_orig(orig):
        if orig.data.node is not None:
            return (
                '<i>{}</i>'.format(html_render_node(orig.data.node)),
            )
        return ()

    def print_result_builder(name):
        return lambda value: (name, str(value))

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(res_graph, [
            dot_printer.DataPrinter('___orig', print_orig)
        ] + [
            dot_printer.DataPrinter(v.name, print_result_builder(
                "{} &isin;".format(v.name)
            ))
            for v in var_set
        ]))


class AnalysisResults(object):
    """
    Contains the results of the collecting semantics analysis.
    """
    def __init__(self, cfg, semantics, trace_domain, vars_domain,
                 evaluator):
        self.cfg = cfg
        self.semantics = semantics
        self.trace_domain = trace_domain
        self.vars_domain = vars_domain
        self.evaluator = evaluator

    def save_cfg_to_file(self, file_name):
        """
        Prints the control-flow graph as a DOT file to the given file name.
        """
        save_cfg_to(file_name, self.cfg)

    def save_results_to_file(self, file_name):
        """
        Prints the resulting graph as a DOT file to the given file name.
        Displays the state of each variable at each program point.
        """
        build_resulting_graph(
            file_name,
            self.cfg,
            self.semantics,
            self.trace_domain
        )

    def eval_at(self, node, expr):
        """
        Given a program point, evaluates for each program trace available at
        this program point the given expression.
        """
        return {
            trace: self.evaluator.eval(expr, env)
            for trace, env in self.semantics[node].iteritems()
        }


def collect_semantics(prog, model, merge_pred_builder):

    cfg = prog.visit(CFGBuilder())

    var_set = set(visitors.findall(prog, lambda n: isinstance(n, Variable)))
    vars_idx = {v: i for i, v in enumerate(var_set)}
    vars_domain = domains.Product(*(model[v].domain for v in var_set))
    trace_domain = SimpleTraceLattice(cfg.nodes)

    evaluator = ExprEvaluator(model)
    solver = ExprSolver(model)

    widening_counter = KeyCounter()
    widening_delay = 10

    def do_widen(counter):
        # will widen when counter == widen_delay, then narrow
        return counter == widening_delay

    do_stmt = VarTracker(var_set, vars_domain, vars_idx, evaluator, solver)

    lat = domains.Set(
        domains.Product(
            trace_domain,
            vars_domain
        ),
        merge_pred_builder.build(
            trace_domain,
            vars_domain
        ),
        None  # We don't need a top element here.
    )

    def it(states):
        new_states = states.copy()

        for node in cfg.nodes:
            inputs = [new_states[anc] for anc in cfg.ancestors(node)]
            res = (lat.build([(trace_domain.bottom, vars_domain.top)])
                   if len(inputs) == 0 else reduce(lat.join, inputs))

            res = lat.build([
                (
                    trace_domain.join(trace, trace_domain.build([node])),
                    updated_val
                )
                for trace, values in res
                for updated_val in [
                    node.data.node.visit(do_stmt, values)
                    if node.data.node is not None else values
                ]
                if not vars_domain.is_empty(updated_val)
            ])

            if node.data.is_widening_point:
                if do_widen(widening_counter.get_incr(node)):
                    res = lat.update(new_states[node], res, True)

            new_states[node] = res

        return new_states

    result = {n: lat.bottom for n in cfg.nodes}
    last = None
    while result != last:
        last, result = result, it(result)

    formatted_results = {
        node: {
            trace: {
                v: values[vars_idx[v]] for v in var_set
            }
            for trace, values in state
        }
        for node, state in result.iteritems()
    }

    return AnalysisResults(
        cfg,
        formatted_results,
        trace_domain,
        vars_domain,
        evaluator
    )
