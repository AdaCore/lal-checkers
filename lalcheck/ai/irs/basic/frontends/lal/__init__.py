"""
Provides a libadalang frontend for the Basic IR.
"""

import libadalang as lal
from lalcheck.ai.utils import profile
from lalcheck.tools.logger import log_stdout, log

import analysis
import typers
import utils
import time
from codegen import ConvertUniversalTypes, gen_ir


unknown_typer = typers.unknown_typer


class ExtractionContext(object):
    """
    The libadalang-based frontend interface. Provides method for extracting
    IR programs from Ada source files (see extract_programs), as well as
    a default typer for those programs (see default_typer).

    Note: programs extracted using different ExtractionContext are not
    compatible. Also, this extraction context must be kept alive as long
    as the programs parsed with it are intended to be used.
    """
    def __init__(self, lal_ctx=None):
        self.lal_ctx = lal.AnalysisContext() if lal_ctx is None else lal_ctx

        # Get a dummy node, needed to call static properties of libadalang.
        dummy = self.lal_ctx.get_from_buffer(
            "<dummy>", 'package Dummy is end;'
        ).root

        # Find the Character TypeDecl.
        char_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Character"
        )

        float_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Float"
        )

        self.evaluator = utils.ConstExprEvaluator(
            dummy.p_bool_type,
            dummy.p_int_type,
            float_type,
            char_type,
            dummy.p_universal_int_type,
            dummy.p_universal_real_type
        )

        self.type_models = {}
        self.fun_models = {}
        self.subpdata = {}
        self._internal_typer = self.default_typer()

    @staticmethod
    def for_project(project_file, scenario_vars={}):
        return ExtractionContext(lal.AnalysisContext(
            unit_provider=lal.UnitProvider.for_project(
                project_file,
                scenario_vars
            )
        ))

    @staticmethod
    def empty():
        return ExtractionContext()

    def extract_programs_from_file(self, ada_file):
        """
        :param str ada_file: A path to the Ada source file from which to
            extract programs.

        :return: a Basic IR Program for each subprogram body that exists in the
            given source code.

        :rtype: iterable[irt.Program]
        """
        return self._extract_from_unit(self.lal_ctx.get_from_file(ada_file))

    def extract_programs_from_provider(self, name, kind):
        return self._extract_from_unit(self.lal_ctx.get_from_provider(
            name, kind
        ))

    def extract_programs_from_unit(self, unit):
        """
        :param lal.AnalysisUnit unit: The already parsed compilation unit.
        """
        return self._extract_from_unit(unit)

    @profile()
    def use_model(self, name):
        model_unit = self.lal_ctx.get_from_provider(name, "specification")
        with log_stdout('error'):
            for diag in model_unit.diagnostics:
                print('   {}'.format(diag))
                return

        model_ofs = model_unit.root.findall(
            lambda x: (x.is_a(lal.AspectAssoc)
                       and x.f_id.text.lower() == "model_of")
        )

        type_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.TypeDecl, lal.SubtypeDecl)
        }

        fun_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.SubpDecl)
        }

        if len(type_models) + len(fun_models) != len(model_ofs):
            with log_stdout('info'):
                print("warning: detected usage of Model_Of in an unknown "
                      "context.")

        for tdecl, ref in type_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print("warning: Model_Of used on '{}' refers to an unknown"
                          " type.".format(tdecl.f_name.text))
            else:
                self.type_models[ref] = tdecl

        for fdecl, ref in fun_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print(
                        "warning: Model_Of used on '{}' refers to an unknown "
                        "procedure/function.".format(
                            fdecl.f_subp_spec.f_subp_name.text
                        )
                    )
            else:
                self.fun_models[ref] = fdecl

    @profile()
    def _extract_from_unit(self, unit):
        if unit.root is None:
            with log_stdout('error'):
                print('Could not parse {}:'.format(unit.filename))
                for diag in unit.diagnostics:
                    print('   {}'.format(diag))
                    return

        unit.populate_lexical_env()

        subpdata = analysis.traverse_unit(unit.root)
        self.subpdata.update(subpdata)

        progs = []
        for subp, subpuserdata in subpdata.iteritems():
            if subp.is_a(lal.BaseSubpBody):
                start_t = time.clock()
                progs.append(
                    gen_ir(self, subp, self._internal_typer, subpuserdata)
                )
                end_t = time.clock()
                log(
                    'timings',
                    " - Transformation of subprocedure {} took {}s".format(
                        subp.f_subp_spec.f_subp_name.text,
                        end_t - start_t
                    )
                )

        converter = ConvertUniversalTypes(self.evaluator, self._internal_typer)

        for prog in progs:
            prog.visit(converter)

        return progs

    def standard_typer(self):
        """
        :return: A Typer for Ada standard types of programs parsed using
            this extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.standard_typer(self)

    def model_typer(self, inner_typer):
        return typers.model_typer(self, inner_typer)

    def default_typer(self, fallback_typer=None):
        """
        :return: The default Typer for Ada programs parsed using this
            extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.default_typer(self, fallback_typer)
