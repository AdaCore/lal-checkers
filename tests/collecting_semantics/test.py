import lalcheck.irs.basic.frontends.lal as lal2basic
from lalcheck.irs.basic.tools import Models
from lalcheck.interpretations import default_type_interpreter
from checkers.collecting_semantics import (
    collect_semantics,
    MergePredicateBuilder
)


def do(test_name, merge_predicate, idx=None):
    ctx = lal2basic.new_context()

    progs = lal2basic.extract_programs(
        ctx,
        '{}.adb'.format(test_name)
    )

    model_builder = Models(
        lal2basic.default_typer(ctx),
        default_type_interpreter
    )

    model = model_builder.of(*progs)

    collect_semantics(
        progs[0],
        model,
        merge_predicate,
        'output/{}_cfg.dot'.format(test_name),
        'output/{}{}_res.dot'.format(test_name, '' if idx is None else idx)
    )


le_trace_eq_vals = (
    MergePredicateBuilder.Le_Traces |
    MergePredicateBuilder.Eq_Vals
)

do('test_simple_if', MergePredicateBuilder.Always, '_always')
do('test_simple_if', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_if_expr', MergePredicateBuilder.Always, '_always')
do('test_simple_if_expr', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_while', le_trace_eq_vals)

do('test_simple_enum', MergePredicateBuilder.Always, '_always')

do('test_ptr', MergePredicateBuilder.Always, '_always')
do('test_ptr', le_trace_eq_vals, '_le_t_eq_v')
