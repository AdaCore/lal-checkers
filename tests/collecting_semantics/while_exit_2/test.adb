procedure Ex1 is
   from : Integer := 0;
   to : Integer := 100;
   step: Integer := 2;
   tmp: Integer;
begin
   my_loop:
   while from < to loop
      tmp := step;
      while tmp > 0 loop
         from := from + 1;
         exit my_loop when from >= to;
         tmp := tmp - 1;
      end loop;
   end loop my_loop;
end Ex1;
