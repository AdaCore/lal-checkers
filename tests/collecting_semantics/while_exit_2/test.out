{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(-2147483648, 2147483647)"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(2, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 100)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign5": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 1)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume2",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(100, 100)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 0)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume5": [],
    "loop_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_join1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_join1",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_start0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_start1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 2147483647)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(-2147483648, 2147483647)"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(-2147483648, 2147483647)"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(2, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 100)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assign5": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 1)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume2",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(100, 100)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 0)",
          "to": "(100, 100)"
        }
      }
    ],
    "assume5": [],
    "loop_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_join1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_join1",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_start0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "loop_start1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 99)",
          "step": "(2, 2)",
          "tmp": "(0, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "from": "(0, 0)",
          "step": "(2, 2)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(100, 100)"
        }
      }
    ],
    "split_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assume0",
          "assume1",
          "assume3",
          "assume4",
          "loop_join0",
          "loop_start0",
          "loop_start1",
          "read0",
          "split_join0",
          "start0"
        ],
        "values": {
          "from": "(-2147483647, 99)",
          "step": "(2, 2)",
          "tmp": "(1, 2)",
          "to": "(100, 100)"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "from": "(-2147483648, 2147483647)",
          "step": "(-2147483648, 2147483647)",
          "tmp": "(-2147483648, 2147483647)",
          "to": "(-2147483648, 2147483647)"
        }
      }
    ]
  }
}